# Merge and Prepare Intermediate Datasets -- Medications and Patients

This notebook is the data engineering for the creation of the intermediate dataset for the "Medications" and "Patients" merged dataset. 


The script for Medications: 
- reads in the raw data provided by our project partner,
- drops duplicate records for a particular FRDPersonnelID

The script for Patients: 
- drops unneeded columns
- removes duplicate rows,
- Drops NULL rows for Provider Gender and Start Date

Then the script performs a merge of the Medications and Patients datasets with:
- an inner join on PatientId and FRDPersonnelID
- a left outer join on Medications back to Patients to build Orphan rows dataframe
- build lookup reference dataframe from Patients of all Providers
- inner join lookup dataframe with Orphan rows dataframe
- build lookup reference dataframe from Patients of all Patients
- merge lookup dataframe with Orphan rows dataframe on PatientId only
- append Orphan rows dataframe back to Medications+Patients dataframe
- add tenure by month attribute,
- converts categorical data into numeric,
- adds new numeric columns utilizing the the "One Hot Encoding" method, then
- outputs dataframe to a CSV file.

This script does not:
- remove time travelers (records where the provider start date is after the dispatch date)
- remove records with null in the PatientGender column

_Author: Mark Lee_ <br>
_Edited: Michael McRae_


In [1]:
# Function to identify and print easy to understand variable types 
def get_var_category(series):
    unique_count = series.nunique(dropna=False)
    total_count = len(series)
    if pd.api.types.is_numeric_dtype(series):
        return 'Numerical'
    elif pd.api.types.is_datetime64_dtype(series):
        return 'Date'
    elif unique_count == total_count:
        return 'Text (Unique)'
    else:
        return 'Categorical'

def print_categories(df):
    for column_name in df.columns:
        print(column_name, ": ", get_var_category(df[column_name]))

In [2]:
# Import libraries
import numpy as np
import pandas as pd
import os, time

# Setup HTML display
from IPython.core.display import display, HTML
# Notebook cell width adjustment
display(HTML('<style>.container { width:80% !important; }</style>'))

## Read in Medications dataset

In [3]:
# Read in Medications dataset and capture how long it takes
print(os.getcwd())
readStart=time.time()

dfMedications = pd.read_excel(r'../data/01_raw/20210225-ems-raw-v04.xlsx', 
                        sheet_name='Medications',
                        na_values=['NA'])

# Stop the clock and calculate read time
readStop=time.time()
readTime=readStop-readStart
readMin=np.floor(readTime/60)
readSec=np.floor(readTime-(readMin*60))
print("The file was read in {0:.0f} minutes and {1:.0f} seconds.".format(readMin,readSec))

dfMedications.shape

C:\CAPSTONE\ems-analytics\notebooks
The file was read in 0 minutes and 27 seconds.


(63168, 6)

## Read in Patients dataset

In [4]:
# Read in Patients dataset and capture how long it takes
print(os.getcwd())
readStart=time.time()

dfPatients = pd.read_excel(r'../data/01_raw/20210225-ems-raw-v04.xlsx', 
                        sheet_name='Patients', 
                        na_values=['NA'])

# Stop the clock and calculate read time
readStop=time.time()
readTime=readStop-readStart
readMin=np.floor(readTime/60)
readSec=np.floor(readTime-(readMin*60))
print("The file was read in {0:.0f} minutes and {1:.0f} seconds.".format(readMin,readSec))

dfPatients.shape

C:\CAPSTONE\ems-analytics\notebooks
The file was read in 6 minutes and 45 seconds.


(543774, 12)

# Working with Medications dataframe

### Capture Medications record count in variable to track how many rows ultimately get removed

In [5]:
varMedicationsCount = len(dfMedications.index)
print('Medications record count: ', varMedicationsCount)

Medications record count:  63168


### Getting a count of NULL values, but not dropping them yet

In [6]:
# Just getting a count of NULL values, but not dropping them yet
print("\nCount total NaN at each column in a Medications dataFrame : \n")
dfMedications.isnull().sum()


Count total NaN at each column in a Medications dataFrame : 



Dim_Medication_PK                    0
PatientId                            0
Medication_Given_RXCUI_Code          6
Medication_Given_Description         6
FRDPersonnelID                       0
Medication_Administered_Date_Time    1
dtype: int64

### Display data types

In [7]:
print_categories(dfMedications)

Dim_Medication_PK :  Numerical
PatientId :  Numerical
Medication_Given_RXCUI_Code :  Numerical
Medication_Given_Description :  Categorical
FRDPersonnelID :  Categorical
Medication_Administered_Date_Time :  Date


### Drop Medication_Administered_Date_Time column which is not needed for analytics, Dim_Medication_PK acts as a record-level unique identifier

In [8]:
dfMedications = dfMedications.drop(['Medication_Administered_Date_Time'],axis=1)
dfMedications.columns

Index(['Dim_Medication_PK', 'PatientId', 'Medication_Given_RXCUI_Code',
       'Medication_Given_Description', 'FRDPersonnelID'],
      dtype='object')

In [9]:
dfMedications.head()

Dim_Medication_PK  PatientId  Medication_Given_RXCUI_Code  \
0              65740     479862                       7806.0   
1              65744     479876                      26225.0   
2              65749     479916                     125464.0   
3              65750     479920                     237648.0   
4              65758     480011                     125464.0   

  Medication_Given_Description                        FRDPersonnelID  
0                       Oxygen  1D18E8FC-EE92-E211-A596-78E7D18C3D20  
1         Ondansetron (Zofran)  35CA45A3-9C25-4E4D-957F-2EA9E40B5974  
2                Normal saline  F73B4EF0-B70C-4960-9B92-8BF196A30B57  
3          Dextrose 10 % (D10)  C387923A-F613-E411-A585-F305C4522FCB  
4                Normal saline  08CC5399-F492-E211-A596-78E7D18C3D20

In [10]:
dfMedications.loc[dfMedications['FRDPersonnelID'] == '9E5D40DC-5DC6-E411-80C7-001DD8B71D38']

Dim_Medication_PK  PatientId  Medication_Given_RXCUI_Code  \
3860               84682     537112                      26225.0   
3861               84682     537112                      26225.0   
4877              119059     552215                       4337.0   
4878              119059     552215                       4337.0   
6732              127985     577130                       6960.0   
...                  ...        ...                          ...   
62934             442175    1461295                       1191.0   
62935             442176    1461295                      26225.0   
62936             442176    1461295                      26225.0   
62965             442211    1461909                       1191.0   
62966             442211    1461909                       1191.0   

               Medication_Given_Description  \
3860                   Ondansetron (Zofran)   
3861                   Ondansetron (Zofran)   
4877          Fentanyl Citrate (Sublimaze)    
4878          Fentanyl Citrate (Sublimaze)    
6732                     Midazolam (Versed)   
...                                     ...   
62934  Acetylsalicylic Acid (Aspirin, ASA)    
62935                  Ondansetron (Zofran)   
62936                  Ondansetron (Zofran)   
62965  Acetylsalicylic Acid (Aspirin, ASA)    
62966  Acetylsalicylic Acid (Aspirin, ASA)    

                             FRDPersonnelID  
3860   9E5D40DC-5DC6-E411-80C7-001DD8B71D38  
3861   9E5D40DC-5DC6-E411-80C7-001DD8B71D38  
4877   9E5D40DC-5DC6-E411-80C7-001DD8B71D38  
4878   9E5D40DC-5DC6-E411-80C7-001DD8B71D38  
6732   9E5D40DC-5DC6-E411-80C7-001DD8B71D38  
...                                     ...  
62934  9E5D40DC-5DC6-E411-80C7-001DD8B71D38  
62935  9E5D40DC-5DC6-E411-80C7-001DD8B71D38  
62936  9E5D40DC-5DC6-E411-80C7-001DD8B71D38  
62965  9E5D40DC-5DC6-E411-80C7-001DD8B71D38  
62966  9E5D40DC-5DC6-E411-80C7-001DD8B71D38  

[516 rows x 5 columns]

### Drop duplicate Medications rows, these are related to where FRDPersonnelID = 9E5D40DC-5DC6-E411-80C7-001DD8B71D38
### 516 records, deduped expect to drop 258 records

In [11]:
dfMedications = pd.DataFrame.drop_duplicates(dfMedications)
dfMedications.shape

(62910, 5)

### Capture revised Medications record count in variable

In [12]:
varMedicationsCount2 = len(dfMedications.index)
print('Medications revised record count: ', varMedicationsCount2)

Medications revised record count:  62910


### Display duplicate records dropped count

In [13]:
varMedicationsDiff = varMedicationsCount - varMedicationsCount2
print('Duplicate records count: ', varMedicationsDiff)

Duplicate records count:  258


# Working with Patients dataframe

### Capture Patients record count in variable to track how many rows ultimately get removed

In [14]:
varPatientsCount = len(dfPatients.index)
print('Patients record count: ', varPatientsCount)

Patients record count:  543774


In [15]:
print("\nCount total NaN at each column in a Patients dataFrame : \n")
dfPatients.isnull().sum()


Count total NaN at each column in a Patients dataFrame : 



PatientId                     0
FRDPersonnelID                0
Shift                         0
UnitId                        0
FireStation                   0
Battalion                     0
PatientOutcome                2
PatientGender            131378
CrewMemberRoles               0
DispatchTime                  0
FRDPersonnelGender          474
FRDPersonnelStartDate       474
dtype: int64

### Work on Patients dataset first

In [16]:
print_categories(dfPatients)

PatientId :  Numerical
FRDPersonnelID :  Categorical
Shift :  Categorical
UnitId :  Categorical
FireStation :  Numerical
Battalion :  Numerical
PatientOutcome :  Categorical
PatientGender :  Categorical
CrewMemberRoles :  Categorical
DispatchTime :  Date
FRDPersonnelGender :  Categorical
FRDPersonnelStartDate :  Date


## Keeping necessary columns for Focus Questions
```
PatientId
FRDPersonnelID
PatientOutcome
PatientGender
DispatchTime
FRDPersonnelGender
FRDPersonnelStartDate
```

### Create new dataframe with only columns needed for Focus Questions

In [17]:
dfPatientsSub = dfPatients.iloc[:, [0,1,6,7,9,10,11]]
dfPatientsSub.shape

(543774, 7)

In [18]:
dfPatientsSub.head()

PatientId                        FRDPersonnelID  \
0     479838  6ED7C99E-9E01-E211-B5F5-78E7D18CFD3C   
1     479838  F39A55F0-C619-E511-80CA-001DD8B71D38   
2     479838  D6D3C99E-9E01-E211-B5F5-78E7D18CFD3C   
3     479839  71D4C99E-9E01-E211-B5F5-78E7D18CFD3C   
4     479839  F9D4C99E-9E01-E211-B5F5-78E7D18CFD3C   

                    PatientOutcome PatientGender        DispatchTime  \
0  Standby (No Services Performed)           NaN 2018-01-01 00:29:02   
1  Standby (No Services Performed)           NaN 2018-01-01 00:29:02   
2  Standby (No Services Performed)           NaN 2018-01-01 00:29:02   
3  No Treatment/Transport Required           NaN 2018-01-01 00:40:54   
4  No Treatment/Transport Required           NaN 2018-01-01 00:40:54   

  FRDPersonnelGender FRDPersonnelStartDate  
0               Male            1997-08-18  
1               Male            2007-02-19  
2               Male            2008-08-18  
3               Male            2011-10-24  
4               Male            2012-02-27

### Remove duplicate records 

In [19]:
dfPatientsSub = pd.DataFrame.drop_duplicates(dfPatientsSub)
dfPatientsSub.shape

(543718, 7)

### Capture revised Patients record count in variable

In [20]:
varPatientsSubCount = len(dfPatientsSub.index)
print('Patients record count: ', varPatientsSubCount)

Patients record count:  543718


### Display duplicate records dropped count

In [21]:
varPatientsSubDiff = varPatientsCount - varPatientsSubCount
print('Duplicate records count: ', varPatientsSubDiff)

Duplicate records count:  56


### Just getting a count of NULL values, but not dropping them yet

In [22]:
print("\nCount total NaN at each column in a Medications Sub dataFrame : \n")
dfPatientsSub.isnull().sum()


Count total NaN at each column in a Medications Sub dataFrame : 



PatientId                     0
FRDPersonnelID                0
PatientOutcome                2
PatientGender            131365
DispatchTime                  0
FRDPersonnelGender          460
FRDPersonnelStartDate       460
dtype: int64

### Drop NULL PatientGender records

In [23]:
# Don't run yet

#dfPatientsSub = dfPatientsSub.drop(dfPatientsSub[(dfPatientsSub.PatientGender.isnull())].index)
#dfPatientsSub.shape

### Drop NULL rows for Provider Gender and Start Date (FRDPersonnelGender and FRDPersonnelStartDate are both 460)

In [24]:
dfPatientsSub = dfPatientsSub.drop(dfPatientsSub[(dfPatientsSub.FRDPersonnelGender.isnull())].index)
dfPatientsSub.shape

(543258, 7)

### Capture revised Patients record count in variable

In [25]:
varPatientsSubCount2 = len(dfPatientsSub.index)
print('Patients record count: ', varPatientsSubCount2)

Patients record count:  543258


### Display records dropped count

In [26]:
varPatientsSubDiff2 = varPatientsSubCount - varPatientsSubCount2
print('Null Provider Gender and Start Date Dropped records count: ', varPatientsSubDiff2)

Null Provider Gender and Start Date Dropped records count:  460


# Merge Medications and Patients datasets

Start with an inner join Between the Medications and Patients dataframes on `PatientId` and `FRDPersonnelID`

In [27]:
dfMedPat = dfPatientsSub.merge(dfMedications, 
                        on=('PatientId','FRDPersonnelID'))
dfMedPat.shape

(58699, 10)

In [28]:
dfMedPat.head()

PatientId                        FRDPersonnelID          PatientOutcome  \
0     479862  1D18E8FC-EE92-E211-A596-78E7D18C3D20  Treated & Transported    
1     479876  35CA45A3-9C25-4E4D-957F-2EA9E40B5974  Treated & Transported    
2     479916  F73B4EF0-B70C-4960-9B92-8BF196A30B57  Treated & Transported    
3     480011  08CC5399-F492-E211-A596-78E7D18C3D20  Treated & Transported    
4     480022  E10FD6B2-19A1-468A-825B-7EE0061DBB01  Treated & Transported    

  PatientGender        DispatchTime FRDPersonnelGender FRDPersonnelStartDate  \
0        Female 2018-01-01 00:44:31             Female            2012-09-24   
1          Male 2018-01-01 00:42:53               Male            2016-02-08   
2          Male 2018-01-01 01:48:31               Male            2016-02-08   
3          Male 2018-01-01 03:02:53               Male            2012-09-24   
4        Female 2018-01-01 03:04:20               Male            2016-09-19   

   Dim_Medication_PK  Medication_Given_RXCUI_Code Medication_Given_Description  
0              65740                       7806.0                       Oxygen  
1              65744                      26225.0         Ondansetron (Zofran)  
2              65749                     125464.0                Normal saline  
3              65758                     125464.0                Normal saline  
4              65759                      26225.0         Ondansetron (Zofran)

### Display count of NULLs by column

In [29]:
print("\nCount total NaN at each column in a Medications+Patients dataFrame : \n")
dfMedPat.isnull().sum()


Count total NaN at each column in a Medications+Patients dataFrame : 



PatientId                        0
FRDPersonnelID                   0
PatientOutcome                   0
PatientGender                   32
DispatchTime                     0
FRDPersonnelGender               0
FRDPersonnelStartDate            0
Dim_Medication_PK                0
Medication_Given_RXCUI_Code      6
Medication_Given_Description     6
dtype: int64

## Perform a left outer join on Medications back to Medications+Patients where NULL

### Add Composite Key using PatientId and FRDPersonnelID to both Medications and Medications+Patients datasets for left outer joins

In [30]:
dfMedications['PK'] = dfMedications['PatientId'].map(str) + '_' + dfMedications['FRDPersonnelID']
dfMedPat['_PK'] = dfMedPat['PatientId'].map(str) + '_' + dfMedPat['FRDPersonnelID']

### Check columns

In [31]:
dfMedPat.columns

Index(['PatientId', 'FRDPersonnelID', 'PatientOutcome', 'PatientGender',
       'DispatchTime', 'FRDPersonnelGender', 'FRDPersonnelStartDate',
       'Dim_Medication_PK', 'Medication_Given_RXCUI_Code',
       'Medication_Given_Description', '_PK'],
      dtype='object')

### Display count of NULLs by column

In [32]:
print("\nCount total NaN at each column in a Medications+Patients dataFrame : \n")
dfMedPat.isnull().sum()


Count total NaN at each column in a Medications+Patients dataFrame : 



PatientId                        0
FRDPersonnelID                   0
PatientOutcome                   0
PatientGender                   32
DispatchTime                     0
FRDPersonnelGender               0
FRDPersonnelStartDate            0
Dim_Medication_PK                0
Medication_Given_RXCUI_Code      6
Medication_Given_Description     6
_PK                              0
dtype: int64

# Perform left outer join on Composite Keys (PK and _PK) and create a new dataframe to append to exising dfMedPat

In [33]:
medPKi = dfMedications.set_index('PK')
medPatPKi = dfMedPat.set_index('_PK')

dfMedLeftJoin = dfMedications.merge(medPatPKi, 
                             how = 'left', 
                             left_on = ('PK'),
                             right_on = ('_PK'), 
                             indicator = 'i').query('i == "left_only"')

In [34]:
dfMedLeftJoin.shape

(4211, 17)

In [35]:
dfMedLeftJoin.head()

Dim_Medication_PK_x  PatientId_x  Medication_Given_RXCUI_Code_x  \
3                  65750       479920                       237648.0   
8                  65770       480042                         4337.0   
70                 65887       480388                        26225.0   
123                65978       480637                         1191.0   
135                66016       480751                        26225.0   

           Medication_Given_Description_x  \
3                     Dextrose 10 % (D10)   
8           Fentanyl Citrate (Sublimaze)    
70                   Ondansetron (Zofran)   
123  Acetylsalicylic Acid (Aspirin, ASA)    
135                  Ondansetron (Zofran)   

                         FRDPersonnelID_x  \
3    C387923A-F613-E411-A585-F305C4522FCB   
8    F1D2C99E-9E01-E211-B5F5-78E7D18CFD3C   
70   38D8C99E-9E01-E211-B5F5-78E7D18CFD3C   
123  1FD58791-CD19-E511-80CA-001DD8B71D38   
135  38D8C99E-9E01-E211-B5F5-78E7D18CFD3C   

                                              PK  PatientId_y  \
3    479920_C387923A-F613-E411-A585-F305C4522FCB          NaN   
8    480042_F1D2C99E-9E01-E211-B5F5-78E7D18CFD3C          NaN   
70   480388_38D8C99E-9E01-E211-B5F5-78E7D18CFD3C          NaN   
123  480637_1FD58791-CD19-E511-80CA-001DD8B71D38          NaN   
135  480751_38D8C99E-9E01-E211-B5F5-78E7D18CFD3C          NaN   

    FRDPersonnelID_y PatientOutcome PatientGender DispatchTime  \
3                NaN            NaN           NaN          NaT   
8                NaN            NaN           NaN          NaT   
70               NaN            NaN           NaN          NaT   
123              NaN            NaN           NaN          NaT   
135              NaN            NaN           NaN          NaT   

    FRDPersonnelGender FRDPersonnelStartDate  Dim_Medication_PK_y  \
3                  NaN                   NaT                  NaN   
8                  NaN                   NaT                  NaN   
70                 NaN                   NaT                  NaN   
123                NaN                   NaT                  NaN   
135                NaN                   NaT                  NaN   

     Medication_Given_RXCUI_Code_y Medication_Given_Description_y          i  
3                              NaN                            NaN  left_only  
8                              NaN                            NaN  left_only  
70                             NaN                            NaN  left_only  
123                            NaN                            NaN  left_only  
135                            NaN                            NaN  left_only

### Display count of NULLs by column

In [36]:
print("\nCount total NaN at each column in a Medications+Patients dataFrame : \n")
dfMedLeftJoin.isnull().sum()


Count total NaN at each column in a Medications+Patients dataFrame : 



Dim_Medication_PK_x                  0
PatientId_x                          0
Medication_Given_RXCUI_Code_x        0
Medication_Given_Description_x       0
FRDPersonnelID_x                     0
PK                                   0
PatientId_y                       4211
FRDPersonnelID_y                  4211
PatientOutcome                    4211
PatientGender                     4211
DispatchTime                      4211
FRDPersonnelGender                4211
FRDPersonnelStartDate             4211
Dim_Medication_PK_y               4211
Medication_Given_RXCUI_Code_y     4211
Medication_Given_Description_y    4211
i                                    0
dtype: int64

### Rename columns with _x

In [37]:
dfMedLeftJoin.rename(columns = {'Dim_Medication_PK_x':'Dim_Medication_PK',
                                'PatientId_x':'PatientId',
                                'Medication_Given_RXCUI_Code_x':'Medication_Given_RXCUI_Code',
                                'Medication_Given_Description_x':'Medication_Given_Description',
                                'FRDPersonnelID_x':'FRDPersonnelID'}, inplace = True) 
dfMedLeftJoin.columns

Index(['Dim_Medication_PK', 'PatientId', 'Medication_Given_RXCUI_Code',
       'Medication_Given_Description', 'FRDPersonnelID', 'PK', 'PatientId_y',
       'FRDPersonnelID_y', 'PatientOutcome', 'PatientGender', 'DispatchTime',
       'FRDPersonnelGender', 'FRDPersonnelStartDate', 'Dim_Medication_PK_y',
       'Medication_Given_RXCUI_Code_y', 'Medication_Given_Description_y', 'i'],
      dtype='object')

### Drop all NaN, _y columns, don't need them anymore

In [38]:
dfMedLeftJoin = dfMedLeftJoin.drop(['PatientId_y', 'FRDPersonnelID_y', 'PatientOutcome', 'PatientGender',
                                    'DispatchTime', 'FRDPersonnelGender', 'FRDPersonnelStartDate',
                                    'Dim_Medication_PK_y', 'Medication_Given_RXCUI_Code_y', 
                                    'Medication_Given_Description_y', 'i'],axis=1)
dfMedLeftJoin.columns

Index(['Dim_Medication_PK', 'PatientId', 'Medication_Given_RXCUI_Code',
       'Medication_Given_Description', 'FRDPersonnelID', 'PK'],
      dtype='object')

In [39]:
dfMedLeftJoin.shape

(4211, 6)

### Display count of NULLs by column

In [40]:
print(" \nCount total NaN at each column in a Medications+Patients dataFrame : \n\n", dfMedLeftJoin.isnull().sum())

 
Count total NaN at each column in a Medications+Patients dataFrame : 

 Dim_Medication_PK               0
PatientId                       0
Medication_Given_RXCUI_Code     0
Medication_Given_Description    0
FRDPersonnelID                  0
PK                              0
dtype: int64


## Build lookup reference dataframe from Patients dataframe containing a distinct list of 
```
FRDPersonnelID
FRDPersonnelGender
FRDPersonnelStartDate
```
## Then join lookup dataframe luProvider with dfMedLeftJoin on PatientId only

In [41]:
# Create Lookup table for Providers
luProvider = dfPatients.groupby(['FRDPersonnelID','FRDPersonnelGender', 'FRDPersonnelStartDate']).size().reset_index(name='count')
luProvider.shape

(1830, 4)

In [42]:
luProvider.head(5)

FRDPersonnelID FRDPersonnelGender  \
0  00D5C99E-9E01-E211-B5F5-78E7D18CFD3C               Male   
1  00D6C99E-9E01-E211-B5F5-78E7D18CFD3C               Male   
2  00D7C99E-9E01-E211-B5F5-78E7D18CFD3C               Male   
3  00D8C99E-9E01-E211-B5F5-78E7D18CFD3C               Male   
4  00D9C99E-9E01-E211-B5F5-78E7D18CFD3C               Male   

  FRDPersonnelStartDate  count  
0            2008-04-28   1200  
1            2007-09-17    388  
2            1993-09-20     21  
3            1991-06-03      5  
4            2003-04-19    249

In [43]:
# Drop count column, not needed
luProvider = luProvider.drop(['count'],axis=1)

# Join lookup dataframe luProvider with dfMedLeftJoin on FRDPersonnelID only

In [44]:
dfMedPatOrphans = dfMedLeftJoin.merge(luProvider, 
                        on=('FRDPersonnelID'))
dfMedPatOrphans.shape

(4208, 8)

In [45]:
dfMedPatOrphans.head()

Dim_Medication_PK  PatientId  Medication_Given_RXCUI_Code  \
0              65750     479920                     237648.0   
1             171166     607195                       4337.0   
2             330909     625060                        296.0   
3             330910     625060                        296.0   
4             347351     674487                     125464.0   

    Medication_Given_Description                        FRDPersonnelID  \
0            Dextrose 10 % (D10)  C387923A-F613-E411-A585-F305C4522FCB   
1  Fentanyl Citrate (Sublimaze)   C387923A-F613-E411-A585-F305C4522FCB   
2          Adenosine (Adenocard)  C387923A-F613-E411-A585-F305C4522FCB   
3          Adenosine (Adenocard)  C387923A-F613-E411-A585-F305C4522FCB   
4                  Normal saline  C387923A-F613-E411-A585-F305C4522FCB   

                                            PK FRDPersonnelGender  \
0  479920_C387923A-F613-E411-A585-F305C4522FCB               Male   
1  607195_C387923A-F613-E411-A585-F305C4522FCB               Male   
2  625060_C387923A-F613-E411-A585-F305C4522FCB               Male   
3  625060_C387923A-F613-E411-A585-F305C4522FCB               Male   
4  674487_C387923A-F613-E411-A585-F305C4522FCB               Male   

  FRDPersonnelStartDate  
0            2014-03-24  
1            2014-03-24  
2            2014-03-24  
3            2014-03-24  
4            2014-03-24

### The count of dfMedPatOrphans is 3 records difference, so quick analysis to see why dfMedLeftJoin = 4211, dfMedPatOrphans = 4208, diff 3

In [46]:
dfMedPatOrphansAnalysis = dfMedPatOrphans

In [47]:
dfMedPatOrphansAnalysis.rename(columns = {'PK':'_PK'}, inplace = True)

In [48]:
medPKi = dfMedLeftJoin.set_index('PK')
medPatPKi = dfMedPatOrphansAnalysis.set_index('_PK')

dfMedPatOrphansAnalysis = dfMedLeftJoin.merge(medPatPKi, 
                             how = 'left', 
                             left_on = ('PK'),
                             right_on = ('_PK'), 
                             indicator = 'i').query('i == "left_only"')

In [49]:
dfMedPatOrphansAnalysis.shape

(3, 14)

In [50]:
dfMedPatOrphansAnalysis

Dim_Medication_PK_x  PatientId_x  Medication_Given_RXCUI_Code_x  \
375                83144       531908                         7213.0   
376                83145       531908                          435.0   
377                83146       531908                         6902.0   

        Medication_Given_Description_x                      FRDPersonnelID_x  \
375      Ipratropium Bromide(Atrovent)  6AD7C99E-9E01-E211-B5F5-78E7D18CFD3C   
376              Albuterol (Ventolin)   6AD7C99E-9E01-E211-B5F5-78E7D18CFD3C   
377  Methylprednisolone (Solu-Medrol)   6AD7C99E-9E01-E211-B5F5-78E7D18CFD3C   

                                              PK  Dim_Medication_PK_y  \
375  531908_6AD7C99E-9E01-E211-B5F5-78E7D18CFD3C                  NaN   
376  531908_6AD7C99E-9E01-E211-B5F5-78E7D18CFD3C                  NaN   
377  531908_6AD7C99E-9E01-E211-B5F5-78E7D18CFD3C                  NaN   

     PatientId_y  Medication_Given_RXCUI_Code_y  \
375          NaN                            NaN   
376          NaN                            NaN   
377          NaN                            NaN   

    Medication_Given_Description_y FRDPersonnelID_y FRDPersonnelGender  \
375                            NaN              NaN                NaN   
376                            NaN              NaN                NaN   
377                            NaN              NaN                NaN   

    FRDPersonnelStartDate          i  
375                   NaT  left_only  
376                   NaT  left_only  
377                   NaT  left_only

### These 3 Provider IDs are not found at all in the Patients dataframe (or the Provider lookup dataframe)

In [51]:
dfMedPatOrphansAnalysis.groupby(['FRDPersonnelID_x']).size().reset_index(name='count')

FRDPersonnelID_x  count
0  6AD7C99E-9E01-E211-B5F5-78E7D18CFD3C      3

# Join dfMedPatOrphans dataframe back with dfPatients to bring back in
```
PatientOutcome
PatientGender
```

In [52]:
# A quick reminder of dfPatients schema
dfPatients.columns

Index(['PatientId', 'FRDPersonnelID', 'Shift', 'UnitId', 'FireStation',
       'Battalion', 'PatientOutcome', 'PatientGender', 'CrewMemberRoles',
       'DispatchTime', 'FRDPersonnelGender', 'FRDPersonnelStartDate'],
      dtype='object')

In [53]:
# A quick reminder of dfMedPatOrphans schema
dfMedPatOrphans.columns

Index(['Dim_Medication_PK', 'PatientId', 'Medication_Given_RXCUI_Code',
       'Medication_Given_Description', 'FRDPersonnelID', '_PK',
       'FRDPersonnelGender', 'FRDPersonnelStartDate'],
      dtype='object')

## Build lookup reference dataframe from Patients dataframe containing a distinct list of 
```
PatientId
PatientOutcome
PatientGender
DispatchTime
```
## Then join/merge lookup dataframe luPatients with dfMedPatOrphans on PatientId only

In [54]:
luPatients = dfPatients.groupby(['PatientId', 'PatientOutcome', 'PatientGender','DispatchTime']).size().reset_index(name='count')
luPatients.shape

(189513, 5)

In [55]:
luPatients.dtypes

PatientId                  int64
PatientOutcome            object
PatientGender             object
DispatchTime      datetime64[ns]
count                      int64
dtype: object

In [56]:
luPatients.head()

PatientId          PatientOutcome PatientGender        DispatchTime  count
0     479851  Patient Refusal  (AMA)          Male 2018-01-01 00:20:49      2
1     479853  Treated & Transported           Male 2018-01-01 00:04:06      2
2     479862  Treated & Transported         Female 2018-01-01 00:44:31      4
3     479876  Treated & Transported           Male 2018-01-01 00:42:53      3
4     479879  Treated & Transported         Female 2018-01-01 01:17:54      2

In [57]:
# Drop count column, not needed
luPatients = luPatients.drop(['count'],axis=1)

In [58]:
dfMedPatSub = dfMedPatOrphans.merge(luPatients,on=('PatientId'))
dfMedPatSub.shape

(4182, 11)

In [59]:
dfMedPatSub.head()

Dim_Medication_PK  PatientId  Medication_Given_RXCUI_Code  \
0              65750     479920                     237648.0   
1             171166     607195                       4337.0   
2             330909     625060                        296.0   
3             330910     625060                        296.0   
4             347351     674487                     125464.0   

    Medication_Given_Description                        FRDPersonnelID  \
0            Dextrose 10 % (D10)  C387923A-F613-E411-A585-F305C4522FCB   
1  Fentanyl Citrate (Sublimaze)   C387923A-F613-E411-A585-F305C4522FCB   
2          Adenosine (Adenocard)  C387923A-F613-E411-A585-F305C4522FCB   
3          Adenosine (Adenocard)  C387923A-F613-E411-A585-F305C4522FCB   
4                  Normal saline  C387923A-F613-E411-A585-F305C4522FCB   

                                           _PK FRDPersonnelGender  \
0  479920_C387923A-F613-E411-A585-F305C4522FCB               Male   
1  607195_C387923A-F613-E411-A585-F305C4522FCB               Male   
2  625060_C387923A-F613-E411-A585-F305C4522FCB               Male   
3  625060_C387923A-F613-E411-A585-F305C4522FCB               Male   
4  674487_C387923A-F613-E411-A585-F305C4522FCB               Male   

  FRDPersonnelStartDate          PatientOutcome PatientGender  \
0            2014-03-24  Treated & Transported         Female   
1            2014-03-24  Treated & Transported         Female   
2            2014-03-24  Treated & Transported           Male   
3            2014-03-24  Treated & Transported           Male   
4            2014-03-24  Treated & Transported           Male   

         DispatchTime  
0 2018-01-01 01:39:54  
1 2018-05-24 17:54:06  
2 2018-06-13 08:18:44  
3 2018-06-13 08:18:44  
4 2018-08-06 16:45:16

In [60]:
dfMedPatSub.columns

Index(['Dim_Medication_PK', 'PatientId', 'Medication_Given_RXCUI_Code',
       'Medication_Given_Description', 'FRDPersonnelID', '_PK',
       'FRDPersonnelGender', 'FRDPersonnelStartDate', 'PatientOutcome',
       'PatientGender', 'DispatchTime'],
      dtype='object')

In [61]:
print("\nCount total NaN at each column in a Medications dataFrame : \n")
dfMedPatSub.isnull().sum()


Count total NaN at each column in a Medications dataFrame : 



Dim_Medication_PK               0
PatientId                       0
Medication_Given_RXCUI_Code     0
Medication_Given_Description    0
FRDPersonnelID                  0
_PK                             0
FRDPersonnelGender              0
FRDPersonnelStartDate           0
PatientOutcome                  0
PatientGender                   0
DispatchTime                    0
dtype: int64

In [62]:
print("\nCount total NaN at each column in a Medications dataFrame : \n")
dfMedPat.isnull().sum()


Count total NaN at each column in a Medications dataFrame : 



PatientId                        0
FRDPersonnelID                   0
PatientOutcome                   0
PatientGender                   32
DispatchTime                     0
FRDPersonnelGender               0
FRDPersonnelStartDate            0
Dim_Medication_PK                0
Medication_Given_RXCUI_Code      6
Medication_Given_Description     6
_PK                              0
dtype: int64

# Append dfMedPatSub back to dfMedPat and do not reindex

In [63]:
dfMedPatAppend = dfMedPat.append(dfMedPatSub, ignore_index=True)
dfMedPatAppend.shape

(62881, 11)

In [64]:
print("\nCount total NaN at each column in a Medications dataFrame : \n")
dfMedPatAppend.isnull().sum()


Count total NaN at each column in a Medications dataFrame : 



PatientId                        0
FRDPersonnelID                   0
PatientOutcome                   0
PatientGender                   32
DispatchTime                     0
FRDPersonnelGender               0
FRDPersonnelStartDate            0
Dim_Medication_PK                0
Medication_Given_RXCUI_Code      6
Medication_Given_Description     6
_PK                              0
dtype: int64

In [65]:
print_categories(dfMedPatAppend)

PatientId :  Numerical
FRDPersonnelID :  Categorical
PatientOutcome :  Categorical
PatientGender :  Categorical
DispatchTime :  Date
FRDPersonnelGender :  Categorical
FRDPersonnelStartDate :  Date
Dim_Medication_PK :  Numerical
Medication_Given_RXCUI_Code :  Numerical
Medication_Given_Description :  Categorical
_PK :  Categorical


In [66]:
dfMedPatAppend['PatientOutcome'].value_counts()

Treated & Transported                        59537
Patient Dead at Scene (EMS CPR Attempted)     1855
Patient Refusal  (AMA)                        1343
Treated, Transferred Care                      112
EMS Assist (Other Agency)                       32
Patient Dead at Scene (No EMS CPR)               2
Name: PatientOutcome, dtype: int64

In [67]:
dfMedPatAppend['PatientGender'].value_counts()

Male                             32536
Female                           30289
Unknown (Unable to Determine)       24
Name: PatientGender, dtype: int64

In [68]:
dfMedPatAppend['FRDPersonnelGender'].value_counts()

Male      52356
Female    10525
Name: FRDPersonnelGender, dtype: int64

In [69]:
dfMedPatAppend['Medication_Given_Description'].value_counts()

Oxygen                                  10707
Normal saline                            7552
Ondansetron (Zofran)                     6855
Fentanyl Citrate (Sublimaze)             6577
Epinephrine 0.1 MG/ML (1:10,000)         6189
Acetylsalicylic Acid (Aspirin, ASA)      4349
Albuterol (Ventolin)                     3837
Ipratropium Bromide(Atrovent)            2817
Nitroglycerin                            2721
Naloxone (Narcan)                        2056
Midazolam (Versed)                       1395
Dextrose 10 % (D10)                      1187
Glucose (Oral)                            999
Methylprednisolone (Solu-Medrol)          981
Diphenhydramine (Benadryl)                915
Dextrose 50 % (D50)                       866
Ketamine (Ketalar)                        642
Epinephrine 1 MG/ML (1:1,000)             565
Adenosine (Adenocard)                     393
Atropine (Atropine Sulfate)               313
Amiodarone (Cordarone)                    294
Glucagon (Glucagen)               

# Add TenureMonths, which is the count of months working at the time of the dispatch call

In [70]:
dfMedPatAppend['TenureMonths'] = ((dfMedPatAppend.loc[:, 'DispatchTime'].dt.date - \
                                     dfMedPatAppend.loc[:, 'FRDPersonnelStartDate'].dt.date) / \
                                    np.timedelta64(1, 'M')).astype(int)
dfMedPatAppend.shape
# error is just a warning, need to figure out even though using loc

(62881, 12)

In [71]:
dfMedPatAppend['TenureMonths'].value_counts()

 55     658
 51     656
 49     650
 59     629
 56     626
       ... 
-20       1
 437      1
-11       1
 3        1
 420      1
Name: TenureMonths, Length: 420, dtype: int64

### Add a factorized column starting with 1 as opposed to the normal python 0

In [72]:
# First just going to to add a factorized column starting with 1 as opposed to the normal python 0
dfMedPatAppend['PatientOutcomeCode'] = pd.factorize(dfMedPatAppend['PatientOutcome'])[0] + 1

In [73]:
dfMedPatAppend['PatientOutcomeCode'].value_counts()

1    59537
4     1855
3     1343
2      112
5       32
6        2
Name: PatientOutcomeCode, dtype: int64

In [74]:
dfMedPatAppend['PatientGenderCode'] = pd.factorize(dfMedPatAppend['PatientGender'])[0] + 1

In [75]:
dfMedPatAppend['PatientGenderCode'].value_counts()

2    32536
1    30289
0       32
3       24
Name: PatientGenderCode, dtype: int64

### Use the "One Hot Encoding" method using get_dummies to create a row of 1,0 for each PatientGender

In [76]:
dumDumPatient = pd.get_dummies(dfMedPatAppend['PatientGender'], prefix='PatientGender',)

In [77]:
dumDumPatient.head()

PatientGender_Female  PatientGender_Male  \
0                     1                   0   
1                     0                   1   
2                     0                   1   
3                     0                   1   
4                     1                   0   

   PatientGender_Unknown (Unable to Determine)  
0                                            0  
1                                            0  
2                                            0  
3                                            0  
4                                            0

In [78]:
dumDumPatient.value_counts()

PatientGender_Female  PatientGender_Male  PatientGender_Unknown (Unable to Determine)
0                     1                   0                                              32536
1                     0                   0                                              30289
0                     0                   0                                                 32
                                          1                                                 24
dtype: int64

In [79]:
dfMedPatAppend = pd.concat([dfMedPatAppend, dumDumPatient], axis=1)

In [80]:
dfMedPatAppend.head()

PatientId                        FRDPersonnelID          PatientOutcome  \
0     479862  1D18E8FC-EE92-E211-A596-78E7D18C3D20  Treated & Transported    
1     479876  35CA45A3-9C25-4E4D-957F-2EA9E40B5974  Treated & Transported    
2     479916  F73B4EF0-B70C-4960-9B92-8BF196A30B57  Treated & Transported    
3     480011  08CC5399-F492-E211-A596-78E7D18C3D20  Treated & Transported    
4     480022  E10FD6B2-19A1-468A-825B-7EE0061DBB01  Treated & Transported    

  PatientGender        DispatchTime FRDPersonnelGender FRDPersonnelStartDate  \
0        Female 2018-01-01 00:44:31             Female            2012-09-24   
1          Male 2018-01-01 00:42:53               Male            2016-02-08   
2          Male 2018-01-01 01:48:31               Male            2016-02-08   
3          Male 2018-01-01 03:02:53               Male            2012-09-24   
4        Female 2018-01-01 03:04:20               Male            2016-09-19   

   Dim_Medication_PK  Medication_Given_RXCUI_Code  \
0              65740                       7806.0   
1              65744                      26225.0   
2              65749                     125464.0   
3              65758                     125464.0   
4              65759                      26225.0   

  Medication_Given_Description                                          _PK  \
0                       Oxygen  479862_1D18E8FC-EE92-E211-A596-78E7D18C3D20   
1         Ondansetron (Zofran)  479876_35CA45A3-9C25-4E4D-957F-2EA9E40B5974   
2                Normal saline  479916_F73B4EF0-B70C-4960-9B92-8BF196A30B57   
3                Normal saline  480011_08CC5399-F492-E211-A596-78E7D18C3D20   
4         Ondansetron (Zofran)  480022_E10FD6B2-19A1-468A-825B-7EE0061DBB01   

   TenureMonths  PatientOutcomeCode  PatientGenderCode  PatientGender_Female  \
0            63                   1                  1                     1   
1            22                   1                  2                     0   
2            22                   1                  2                     0   
3            63                   1                  2                     0   
4            15                   1                  1                     1   

   PatientGender_Male  PatientGender_Unknown (Unable to Determine)  
0                   0                                            0  
1                   1                                            0  
2                   1                                            0  
3                   1                                            0  
4                   0                                            0

### Use the "One Hot Encoding" method using get_dummies to create a row of 1,0 for each Provider Gender

In [81]:
dfMedPatAppend['ProviderGenderCode'] = pd.factorize(dfMedPatAppend['FRDPersonnelGender'])[0] + 1

In [82]:
dfMedPatAppend['ProviderGenderCode'].value_counts()

2    52356
1    10525
Name: ProviderGenderCode, dtype: int64

In [83]:
print_categories(dfMedPatAppend)

PatientId :  Numerical
FRDPersonnelID :  Categorical
PatientOutcome :  Categorical
PatientGender :  Categorical
DispatchTime :  Date
FRDPersonnelGender :  Categorical
FRDPersonnelStartDate :  Date
Dim_Medication_PK :  Numerical
Medication_Given_RXCUI_Code :  Numerical
Medication_Given_Description :  Categorical
_PK :  Categorical
TenureMonths :  Numerical
PatientOutcomeCode :  Numerical
PatientGenderCode :  Numerical
PatientGender_Female :  Numerical
PatientGender_Male :  Numerical
PatientGender_Unknown (Unable to Determine) :  Numerical
ProviderGenderCode :  Numerical


### Output the intemediate data Medications+Patients dataframe to CSV into the Intermediate data directory

In [84]:
dfMedPatAppend.to_csv(r'../data/02_intermediate/MedicationsPatients-Intermediate.csv', index = False)

### There are still some "time travelers" in the data. These should be removed for any tenure analyses.

In [85]:
len(dfMedPatAppend[dfMedPatAppend['TenureMonths']<0])

2

In [86]:
dfMedPatAppend[dfMedPatAppend['TenureMonths']<0]

PatientId                        FRDPersonnelID  \
4207      547276  6FB584B8-F37D-45E3-8FCC-C820ACF46816   
16830     754276  86D2C99E-9E01-E211-B5F5-78E7D18CFD3C   

                                  PatientOutcome PatientGender  \
4207                      Treated & Transported           Male   
16830  Patient Dead at Scene (EMS CPR Attempted)          Male   

             DispatchTime FRDPersonnelGender FRDPersonnelStartDate  \
4207  2018-03-17 10:43:35               Male            2019-12-09   
16830 2018-11-05 18:55:39             Female            2019-10-29   

       Dim_Medication_PK  Medication_Given_RXCUI_Code  \
4207               88238                       7806.0   
16830             376450                       7806.0   

      Medication_Given_Description  \
4207                        Oxygen   
16830                       Oxygen   

                                               _PK  TenureMonths  \
4207   547276_6FB584B8-F37D-45E3-8FCC-C820ACF46816           -20   
16830  754276_86D2C99E-9E01-E211-B5F5-78E7D18CFD3C           -11   

       PatientOutcomeCode  PatientGenderCode  PatientGender_Female  \
4207                    1                  2                     0   
16830                   4                  2                     0   

       PatientGender_Male  PatientGender_Unknown (Unable to Determine)  \
4207                    1                                            0   
16830                   1                                            0   

       ProviderGenderCode  
4207                    2  
16830                   1